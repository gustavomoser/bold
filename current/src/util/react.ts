import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
