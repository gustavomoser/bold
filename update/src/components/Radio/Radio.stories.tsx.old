import { action } from '@storybook/addon-actions'
import { boolean, text } from '@storybook/addon-knobs'
import React from 'react'

import { HFlow } from '../HFlow'

import { Radio } from './Radio'

export default {
  title: 'Components/Radio',
}

export const Default = () => (
  <HFlow>
    <Radio
      name='radio1'
      label={text('label', 'Component label')}
      value='1'
      onChange={action('changed')}
      disabled={boolean('disabled', false)}
    />
    <Radio
      name='radio1'
      label={text('label', 'Component label')}
      value='2'
      onChange={action('changed')}
      disabled={boolean('disabled', false)}
    />
  </HFlow>
)
