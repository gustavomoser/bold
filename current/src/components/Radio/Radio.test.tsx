import { render } from '@testing-library/react'
import React from 'react'

import { Radio } from './Radio'

it('should render correctly', () => {
  const { container } = render(<Radio label='First' />)
  expect(container).toMatchSnapshot()
})
