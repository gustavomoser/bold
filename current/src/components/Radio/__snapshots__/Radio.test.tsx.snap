// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 100px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
}

.emotion-1:after {
  background-color: #FFFFFF;
  border: 5px solid #FFFFFF;
  border-radius: 100px;
  content: "";
  display: block;
  height: 2px;
  margin-left: 6px;
  margin-top: 6px;
  opacity: 0;
  text-align: center;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  width: 2px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  opacity: 1;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-3"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <span
      class="emotion-1"
    />
    <span
      class="emotion-2"
    >
      First
    </span>
  </label>
</div>
`;
